src/game_battler.cpp

```diff
@@ -361,10 +361,36 @@ bool Game_Battler::AddState(int state_id, bool allow_battle_states) {
 	return was_added;
 }
 
+void ForEachStateRemoved(Game_Battler& battler, Span<int16_t> removed_states) {
+	for (const auto& state_id: removed_states) {
```

Comment:
It's not efficient to loop over every state like this, and you don't have to.

Just do:
1. get prev restriction
2. get was_dead
3. apply all state removals (templated lambda which each main function passes in)
4. If was_dead and not is_dead -> Set hp to 1
5. get cur restriction
6. clear battle algo if restriciton changed

Reply:
ah sorry. Now I get it. Also will not use ``IsDead`` but do a state check instead because IsDead checks for HP due to RPG_RT compat.

Reply:
Good catch, yes state check is the right way to go. I would put a comment explaining why in case some developer tries to "fix" it later.

Reply:
Should be correct now.